File: ruby/numbers/e-to-nth-digit.rb
#!/usr/bin/env ruby

def main()
	errorLimit = get_input()
	approxe = calculateE(errorLimit)
	puts "Approximate value of e: #{approxe}"
end

def factorial(num)
	if num < 0
		return -1
	end

	factorial = 1;
	if num == 0
		return factorial
	end

	counter = 1
	while counter <= num
		factorial *= counter
		counter += 1
	end
	return factorial
end

def get_input()
	print "Enter decimals of accuracy desired (no larger than 15): "
	input = gets

	decimals = input.to_i;
	if decimals == 0 || decimals < 0
		puts "Enter a valid, non-zero natural number for the number of decimals to calculate to."
		return 0
	end

	errorLimit = 10 ** (-decimals)
end

def calculateE(errorLimit)
	if errorLimit <= 0
		return 0
	end

	e = 0.0
	index = 0
	error = 2 * errorLimit
	factorial = factorial(0)
	while error > errorLimit
		term = 1.0 / factorial
		e += term

		index += 1
		factorial *= index

		#There is some inaccuracy in using the term as
		#the error itself, but it's not really a big
		#deal.
		error = term
	end
	return e
end

main

File: scala/numbers/e-to-nth-digit.scala
#!/usr/bin/env scala

def calculateE(error_limit: Double): Double = {
	var e = 0.0;
	var index = 0;
	var error = getNthTermOfSeries(index);
	while(error > error_limit) {
		e += error;
		index += 1;
		error = getNthTermOfSeries(index);
	}
	return e;
}

def getNthTermOfSeries(n: Int): Double = {
	val term = 1.0 / factorial(n);
	return term.doubleValue();
}

def factorial(n: Int): Long = {
	if(n == 1 || n == 0) {
		return 1;
	}
	return factorial(n - 1) * n;
}

def main(args: Array[String]): Unit = {
	if(args.length < 1) {
		Console.err.println("Not enough arguments");
		return;
	}
	var decimals = 0;
	try {
		decimals = scala.math.min(Integer.parseInt(args(0)), 15);
	} catch {
		case ex: NumberFormatException => {
			Console.err.println("Unable to parse given argument");
			return;
		}
	}

	if(decimals <= 0) {
		Console.err.println("The number of decimals must be positive");
		return;
	}

	val error_limit = scala.math.pow(10, -decimals);
	val e = calculateE(error_limit);
	println("The value of e is " + e);
}

main(args);
